pkg/runtime/pubsub/subscriptions.go

```diff
@@ -0,0 +1,68 @@
+package pubsub
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+
+	"github.com/dapr/dapr/pkg/channel"
+	"github.com/dapr/dapr/pkg/logger"
+	invokev1 "github.com/dapr/dapr/pkg/messaging/v1"
+	daprclientv1pb "github.com/dapr/dapr/pkg/proto/daprclient/v1"
+	"github.com/golang/protobuf/ptypes/empty"
+)
+
+const (
+	getTopicsError         = "error getting topic list from app: %s"
+	deserializeTopicsError = "error getting topics from app: %s"
+	noSubscriptionsError   = "user app did not subscribe to any topic"
+)
+
+func GetSubscriptionsHTTP(channel channel.AppChannel, log logger.Logger) []Subscription {
+	var subscriptions []Subscription
+	req := invokev1.NewInvokeMethodRequest("dapr/subscribe")
+	req.WithHTTPExtension(http.MethodGet, "")
+	req.WithRawData(nil, invokev1.JSONContentType)
+
+	// TODO Propagate Context
+	ctx := context.Background()
+	resp, err := channel.InvokeMethod(ctx, req)
+	if err != nil {
+		log.Errorf(getTopicsError, err)
+	}
+
+	switch resp.Status().Code {
+	case http.StatusOK:
+		_, body := resp.RawData()
+		if err := json.Unmarshal(body, &subscriptions); err != nil {
+			log.Errorf(deserializeTopicsError, err)
+		}
+	case http.StatusNotFound:
+		log.Debug(noSubscriptionsError)
+
+	default:
+		log.Warnf("app returned http status code %v from subscription endpoint", resp.Status().Code)
+	}
+	return subscriptions
```

Comment:
before returning from here, verify that each topic has a route as route is require, log a warning for topic which doesn't have a route and remove it from list. Else route would be nil for http app channel we will try to invoke on a nil route.

Reply:
I think json.unmarshal will catch the case where route is not given.

Reply:
i think,  json unmarshall should not do that. If it does that, it would be wrong because metadata is optional and would not be in the response body always.

Reply:
I added validation for it.

Reply:
jfyi, json.unmarshal does not catch missing values as errors.

Reply:
I was thinking that json annotation could specify required field. But no.

Good to know :)

Reply:
Does the json annotation has a way to specify a property as required, if yes then unmarshal could have taken care of it.